File: cpp/19_SubarraySumEqualsK.cpp
/*
 * Q19. Subarray Sum Equals K (Category: Sliding Window / Misc)
 *
 * Problem: Return the total number of subarrays whose sum equals k.
 *
 * Approach: Prefix sum + frequency map.
 *           For each index, count how many previous prefix sums equal (prefixSum - k).
 *
 * Time Complexity : O(n)
 * Space Complexity: O(n)
 */
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

int subarraySum(vector<int>& nums, int k) {
    unordered_map<int, int> prefixCount;
    prefixCount[0] = 1;
    int prefixSum = 0, count = 0;

    for (int num : nums) {
        prefixSum += num;
        count += prefixCount[prefixSum - k];
        prefixCount[prefixSum]++;
    }
    return count;
}

int main() {
    vector<int> nums1 = {1, 1, 1};
    cout << subarraySum(nums1, 2) << endl; // 2

    vector<int> nums2 = {1, 2, 3};
    cout << subarraySum(nums2, 3) << endl; // 2
    return 0;
}

File: python/19_subarray_sum_equals_k.py
"""
Q19. Subarray Sum Equals K (Category: Sliding Window / Misc)

Problem: Return the total number of subarrays whose sum equals k.

Approach: Prefix sum + frequency dict.
          For each index, count how many previous prefix sums equal (prefix_sum - k).

Time Complexity : O(n)
Space Complexity: O(n)
"""
from collections import defaultdict


def subarray_sum(nums: list[int], k: int) -> int:
    prefix_count: dict[int, int] = defaultdict(int)
    prefix_count[0] = 1
    prefix_sum = 0
    count = 0

    for num in nums:
        prefix_sum += num
        count += prefix_count[prefix_sum - k]
        prefix_count[prefix_sum] += 1

    return count


if __name__ == "__main__":
    print(subarray_sum([1, 1, 1], 2))  # 2
    print(subarray_sum([1, 2, 3], 3))  # 2
